File: DONGUK/프로그래머스_WEEK_13/징검다리 건너기.py
def solution(stones, k):
    answer = 0
    
    N = len(stones)
    
    start = 0
    end = max(stones)
    
    while start <= end:
        mid = (start + end) // 2
        
        dis = 0
        flag = True
        for i in range(N):
            if stones[i] < mid:
                dis += 1
                
                if dis >= k:
                    break
            
            else:
                if dis >= k:
                    break
                    
                else:
                    dis = 0
            
            
        if dis >= k:
            flag = False
            
        if flag:
            start = mid + 1
            answer = mid
            
        else:
            end = mid - 1
            
    return answer

File: JUNHEE/프로그래머스_옛날문제/징검다리 건너기.cpp
#include <string>
#include <vector>
#include <algorithm>

using namespace std;

int solution(vector<int> stones, int k) {
    int answer = 0;
    int end = *max_element(stones.begin(), stones.end());
    int start = *min_element(stones.begin(), stones.end());
    while(start <= end){
        int mid = (end + start) / 2;
        vector<int> t_stone (stones);
        int max_jump_count = 0;
        int temp_jump_count = 0;
        for(int i = 0; i < stones.size(); i++){
            t_stone[i] -= mid;
            if(t_stone[i] < 0){
                temp_jump_count++;
            }
            else{
                temp_jump_count = 0;
            }
            max_jump_count = max(max_jump_count, temp_jump_count);
        }
        if(max_jump_count < k){
            answer = max(mid, answer);
            start = mid + 1;
        }
        else{
            end = mid - 1;
        }
    }
    
    return answer;
}
